cirq/sim/clifford_simulator.py

```diff
@@ -0,0 +1,749 @@
+                        measurements[k] = []
+                    measurements[k].append(np.array(v, dtype=bool))
+                    
+        return {k: np.array(v) for k, v in measurements.items()}
+
+class CliffordTrialResult(simulator.SimulationTrialResult):
+    def __init__(self,
+            measurements: Dict[str, np.ndarray],
+            final_simulator_state: 'CliffordState') -> None:
+
+        super().__init__(params=None,
+                         measurements=measurements,
+                         final_simulator_state=final_simulator_state)
+
+        self.final_state = final_simulator_state
+        
+    def __str__(self):
+        samples = super().__str__()
+        final = self._final_simulator_state
+        
+        return 'measurements: {}\noutput state: {}'.format(samples, final)
+
+    def __repr__(self):
+        return super().__repr__()
+
+class CliffordSimulatorStepResult(simulator.StepResult):
+    """A `StepResult` that includes `StateVectorMixin` methods."""
+
+    def __init__(self, state, measurements):
+        """Results of a step of the simulator.
+        Attributes:
+            state: A CliffordState
+            measurements: A dictionary from measurement gate key to measurement
+                results, ordered by the qubits that the measurement operates on.
+            qubit_map: A map from the Qubits in the Circuit to the the index
+                of this qubit for a canonical ordering. This canonical ordering
+                is used to define the state vector (see the state_vector()
+                method).
+        """
+        self.measurements = measurements
+        # self.qubit_map = qubit_map
+        self.state = state
+    
+    def __str__(self):
+        def bitstring(vals):
+            return ''.join('1' if v else '0' for v in vals)
+
+        results = sorted(
+            [(key, bitstring(val)) for key, val in self.measurements.items()])
+        
+        if len(results) == 0:
+            measurements = ''
+        else:
+            measurements =  ' '.join(
+                ['{}={}'.format(key, val) for key, val in results]) + '\n'
+        
+        final = self.state
+        
+        return '{}{}'.format(measurements, final)
+    
+    def _simulator_state(self):
+        return self.state
+    
+    def sample(self, qubits: List[ops.Qid],
+               repetitions: int = 1) -> np.ndarray:
+        
+        measurements = []
+        
+        for _ in range(repetitions):
+            measurements.append(self.state.perform_measurement(qubits,collapse_wavefunction=False))
+        
+        return np.array(measurements,dtype=bool)
+
+class CliffordState():
+    """A state of the Clifford simulation. 
+
+    The state is stored using two complementary representations: Anderson's tableaux form
+    and Bravyi's CH-form. The tableaux keeps track of the stabilizer operations, while the
+    CH-form allows access to the full wavefunction (including phase).
```

Comment:
It feels strange to keep track of both, especially for a simulator where we do care about efficiency.  I think you should pass which of these you want into the simulator constructor and then use that to decide which state to use.  You probably will end up with two state classes, which is ok, especially since they support different ways to get information about the simulation out of them.

Reply:
This is a great point, and I agree that decoupling the two representations would be a more elegant solution. However there are two caveats. 

1) I'm not sure how easy it is to perform measurements with Bravyi's representation alone (the current implementation relies on the tableaux for performing measurements). Unless we can figure this out, the Bravyi-based simulation would be limited to unitary operators only. 

2) We can drop the Bravyi representation entirely without too much loss of functionality -- the only thing missing would be the ability to get wavefunction amplitudes in the computational basis. But this functionality is necessary if one wants to improve the efficiency of non-Clifford operations (https://arxiv.org/abs/1808.00128).

Reply:
@dabacon Let's get this in and address decoupling the representations here: https://github.com/quantumlib/Cirq/issues/2123.

Reply:
On second thought, let's put `CH_Form` into a different PR.